### PR TITLE
[feature][task-flink] Support Flink application mode

### LINKED ISSUE #9416: [Feature][task-flink] support deploy a flink task with application mode、Per-Job Cluster Mode、session mode when deploy flink on yarn
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

only yarn-cluster mode was supported now,when deploy a flink task on yarn.but this mode is deprecated when use flink which version is equal or greater   1.12.0.

### Use case

support deploy a flink task with application mode、Per-Job Cluster Mode、session mode when deploy flink on yarn which version is equal or greater   1.12.0.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by paul8263**
## Purpose of the pull request

Added feature that enables Flink to submit jobs in application mode.

This pr will close #9416 

## Brief change log

- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtils.java
- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkConstants.java
- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java
- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkParameters.java
- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkTask.java
- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/test/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtilsTest.java
- dolphinscheduler-ui-next/src/views/projects/task/components/node/fields/use-flink.ts
- dolphinscheduler-ui/src/js/conf/home/pages/dag/_source/formModel/tasks/flink.vue
## Verify this pull request

This change added tests and can be verified as follows:

Added test for dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtils.java
- dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/test/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtilsTest.java

### CODE REVIEW

**Comment by SbloodyS** [comment]
I think your describe see issue ```9416``` is not work for this case.
to target specific issue from PR, we recommend use keyword fix: #issue_id or close: #issue_id or closes: #issue_id in your PR describe(not title, jus desc). I would not only connect issue to PR but also close issue automatically when PR is be closed.

File: dolphinscheduler-ui/src/js/conf/home/pages/dag/_source/formModel/tasks/flink.vue [hunk 1]
```diff
@@ -60,10 +60,11 @@
         <el-radio-group v-model="deployMode" size="small">
           <el-radio :label="'cluster'" :disabled="isDetails"></el-radio>
           <el-radio :label="'local'" :disabled="isDetails"></el-radio>
+          <el-radio :label="'application'" :disabled="isDetails"></el-radio>
```
Comment (SbloodyS) [inline dolphinscheduler-ui/src/js/conf/home/pages/dag/_source/formModel/tasks/flink.vue]: As we discuss in #9416 .```dolphinscheduler-ui``` has been deprecated. This place needs no modification.

Comment (paul8263) [reply dolphinscheduler-ui/src/js/conf/home/pages/dag/_source/formModel/tasks/flink.vue]: Hi @SbloodyS ,
I reverted `flink.vue`.

**Comment by paul8263** [comment]
The docs were updated in https://github.com/apache/dolphinscheduler/commit/fff3d69831ffcf353f2421af15053be47fb7a151

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package org.apache.dolphinscheduler.plugin.task.flink;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Flink deploy mode
+ */
+public enum FlinkDeployMode {
```
Comment (SbloodyS) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java:25]: Would you please kindly add ```yarn-session``` and ```yarn-per-job```? Thanks.

Comment (paul8263) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java:25]: Hi @SbloodyS ，
`local` and `yarn-cluster` is for Flink legacy version, while `yarn-session` and `yarn-per-job` is for newer version.

Legacy Flink:
`flink run [-m yarn-cluster] xxx`
Newer Flink:
`flink run/run-application -t /yarn-per-job/yarn-session/yarn-application`

The `local` mode in newer Flink is session mode and `yarn-cluster` is `yarn-per-job` as well, while yarn-application is brand new.

Newer Flink can also accept legacy deploy mode parameters. So personally I suggest we keep using the legacy ones in order to ensure its compatability.

Correct me if I am wrong.

See [Flink-1.11 Yarn](https://nightlies.apache.org/flink/flink-docs-release-1.11/ops/deployment/yarn_setup.html#run-a-single-flink-job-on-yarn) and [Flink-1.12 Yarn](https://nightlies.apache.org/flink/flink-docs-release-1.12/deployment/resource-providers/yarn.html#per-job-cluster-mode).

Comment (SbloodyS) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java:25]: You are right. Maybe we can refer to the implementation of spark node for spark 2 and spark 3. But this change may be larger. Are you willing to modify it in this PR? It's OK if not. ^_^ @paul8263

Comment (paul8263) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java:25]: Hi @SbloodyS ,
Sorry that I did not quite catch what you mean. Should I implement two kinds of command lines for legacy Flink and the latest one respectively?

Comment (SbloodyS) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java:25]: Yes. Just like the way we do in spark task. @paul8263

Comment (paul8263) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkDeployMode.java:25]: Got it. I will do this.

File: docs/docs/en/guide/task/flink.md [hunk 3]
```diff
@@ -24,7 +24,7 @@ Flink task type for executing Flink programs. For Flink nodes, the worker submit
 - **Program type**: Supports Java, Scala and Python.
 - **The class of main function**: The **full path** of Main Class, the entry point of the Flink program.
 - **Main jar package**: The jar package of the Flink program (upload by Resource Center).
-- **Deployment mode**: Support 2 deployment modes: cluster and local.
+- **Deployment mode**: Support 3 deployment modes: cluster, local and application.
```
Comment (zhongjiajie) [inline docs/docs/en/guide/task/flink.md]: It is better to add an example about how to using flink in `application` mode

Comment (paul8263) [reply docs/docs/en/guide/task/flink.md]: Er.. The only difference is the run command. Should I add some explaination for those 3 modes or simply add a reference link to Flink official website?

Comment (zhongjiajie) [reply docs/docs/en/guide/task/flink.md]: I think adding some links to FL website is better, but please use the fix version docs and avoid using `latest` or `stale` url

**Review by SbloodyS — DISMISSED** [review]
The backend part LGTM. Thanks for your contribution.

**Comment by zhongjiajie** [comment]
@paul8263 have conflict in use-flink.ts, we remove module `dolphinscheduler-ui` and renamed `dolphinscheduler-ui-next` to `dolphinscheduler-ui` in https://github.com/apache/dolphinscheduler/pull/9909, So you have to migrate your change to `dolphinscheduler-ui`

**Comment by zhongjiajie** [comment]
BTW, document change LGTM, we should ask @songjianet to see the frontend change after you solve the conflict

**Comment by paul8263** [comment]
Hi @zhongjiajie ,
Did I miss something? I rebased upstream/dev branch, solved the conflict and then force pushed my working branch. But GitHub says the conflict still exists.

**Comment by paul8263** [comment]
Hi @SbloodyS and @zhongjiajie ,
Got lots of conflicts to resolve. The newly contributed codes seem to be not well structured. It may take some time for me to refactor and solve the conflicts.

**Comment by SbloodyS** [comment]
> Hi @SbloodyS and @zhongjiajie , Got lots of conflicts to resolve. The newly contributed codes seem to be not well structured. It may take some time for me to refactor and solve the conflicts.

Feel free to ping us when you are ready to review. ^_^

**Comment by paul8263** [comment]
Hi @SbloodyS ,
I am confused about getting this error:
java.lang.RuntimeException: No download button in file manage list
	at org.apache.dolphinscheduler.e2e.cases.FileManageE2ETest.testDownloadFile(FileManageE2ETest.java:303)

**Comment by SbloodyS** [comment]
> Hi @SbloodyS , I am confused about getting this error: java.lang.RuntimeException: No download button in file manage list at org.apache.dolphinscheduler.e2e.cases.FileManageE2ETest.testDownloadFile(FileManageE2ETest.java:303)

I had restarted the failed CI.

**Comment by paul8263** [comment]
Hi @SbloodyS ,
Please review this issue as Flink is under development actively. Otherwise we have to frequently rebase and solve the conflicts. Thanks.

**Comment by SbloodyS** [comment]
@labbomb @Amy0104 @devosend PTAL

File: dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-flink.ts [hunk 4]
```diff
@@ -36,14 +36,51 @@ export function useFlink(model: { [field: string]: any }): IJsonItem[] {
...
+      ];
+    }
+  })
+
+  watch(
+    () => model.flinkVersion,
+    () => {
+      model.deployMode = 'cluster'
+    }
+  )
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-flink.ts:85]: There will be a bug here after setting deployMode to application and flinkVersion to 1.11.

Comment (paul8263) [reply dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-flink.ts:85]: Hi @Amy0104 ,
I updated use-flink.ts. Now deploy mode will reset to 'cluster' only when flink version is '<1.10' and deploy mode is 'application'.

**Comment by paul8263** [comment]
@SbloodyS ,
How could I rerun the test suite?

**Comment by paul8263** [comment]
Hi @SbloodyS ,
Could you help review this commit please?

**Review by SbloodyS — DISMISSED** [review]
Backend part LGTM

**Comment by caishunfeng** [comment]
> Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=apache-dolphinscheduler&pullRequest=9577)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=SECURITY_HOTSPOT) [E](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=SECURITY_HOTSPOT) [1 Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=CODE_SMELL) [13 Code Smells](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_coverage&view=list) [60.2% Coverage](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_duplicated_lines_density&view=list)

Hi @paul8263 please take a look of the security hotspots.

**Comment by paul8263** [comment]
> > Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=apache-dolphinscheduler&pullRequest=9577)
> > [Bug](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=SECURITY_HOTSPOT) [E](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=SECURITY_HOTSPOT) [1 Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=CODE_SMELL) [13 Code Smells](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=9577&resolved=false&types=CODE_SMELL)
> > [](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_coverage&view=list) [60.2% Coverage](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=9577&metric=new_duplicated_lines_density&view=list)
> 
> Hi @paul8263 please take a look of the security hotspots.

Hi @caishunfeng ,
Thank you for reviewing. I think the script file permission should be RWXR_XR_X as in other tasks such DataxTask, generated scripts are all granted with the same permission.

**Comment by paul8263** [comment]
Hi @SbloodyS ,
Thanks your for previous suggestions and code reviewing.

Please remind all the reviewers to go over this PR. I am afraid if it was left untouched for a long time, it might diverge greatly from the dev branch. Solving the conflict might be time-consuming.

If there are some places that need to change, please let me know. Thank you very much.

**Comment by SbloodyS** [comment]
@caishunfeng @zhongjiajie @Amy0104 @songjianet @zhuangchong Cound you guys spare some time to take a look at this PR? This PR has been over for 2 months.

**Review by caishunfeng — COMMENTED** [review]
LGTM overall

**Review by Amy0104 — APPROVED** [review]
The front end part LGTM.

**Review by zhuangchong — APPROVED** [review]
+1

**Pull request merged**